src/inference/AdvancedSMC.jl

```diff
@@ -17,76 +43,106 @@ must be stored in a [`TArray`](@ref) object.
 Usage:
 
 ```julia
-SMC(1000)
+SMC()
 ```
 """
-mutable struct SMC{space, F} <: InferenceAlgorithm
-    n_particles           ::  Int
+mutable struct SMC{space, F} <: ParticleInference
     resampler             ::  F
     resampler_threshold   ::  Float64
 end
-function SMC(n_particles::Int, resampler::F, resampler_threshold::Float64, space::Tuple) where {F}
-    return SMC{space, F}(n_particles, resampler, resampler_threshold)
+
+alg_str(spl::Sampler{SMC}) = "SMC"
+function SMC(
+    resampler::F,
+    resampler_threshold::Float64,
+    space::Tuple
+) where {F}
+    return SMC{space, F}(resampler, resampler_threshold)
 end
-SMC(n) = SMC(n, resample_systematic, 0.5, ())
-SMC(n_particles::Int, ::Tuple{}) = SMC(n_particles)
-function SMC(n_particles::Int, space::Symbol...)
-    SMC(n_particles, resample_systematic, 0.5, space)
+SMC() = SMC(resample_systematic, 0.5, ())
+SMC(::Tuple{}) = SMC()
+function SMC(space::Symbol...)
+    SMC(resample_systematic, 0.5, space)
 end
 
-function Sampler(alg::SMC, s::Selector)
-    info = Dict{Symbol, Any}()
-    info[:logevidence] = []
-    return Sampler(alg, info, s)
+mutable struct SMCState <: AbstractSamplerState
+    vi                   ::   TypedVarInfo
+    # The logevidence after aggregating all samples together.
+    average_logevidence  ::   Float64
+    particles            ::   ParticleContainer
 end
 
-function step(model, spl::Sampler{<:SMC}, vi::VarInfo)
+function SMCState(
+    model::M, 
+) where {
+    M<:Model
+}
+    vi = VarInfo(model)
     particles = ParticleContainer{Trace}(model)
-    vi.num_produce = 0;  # Reset num_produce before new sweep\.
-    set_retained_vns_del_by_spl!(vi, spl)
-    resetlogp!(vi)
 
-    push!(particles, spl.alg.n_particles, spl, vi)
+    return SMCState(vi, 0.0, particles)
+end
 
-    while consume(particles) != Val{:done}
-      ess = effectiveSampleSize(particles)
-      if ess <= spl.alg.resampler_threshold * length(particles)
-        resample!(particles,spl.alg.resampler)
-      end
-    end
+function Sampler(alg::T, model::Model, s::Selector) where T<:SMC
+    dict = Dict{Symbol, Any}()
+    state = SMCState(model)
+    return Sampler{T,SMCState}(alg, dict, s, state)
```

Comment:
`{T, SMCState}` seems unnecessary, `Sampler(...)` should work here IIUC. Additionally, when `SMCState` gets type parameters, `SMCState` above will not be a fully concrete type.